src/mpb.cpp

```diff
@@ -267,46 +275,39 @@ void *fields::get_eigenmode(double omega_src,
   if (!quiet && verbose)
    master_printf("KPOINT: %g, %g, %g\n", k[0], k[1], k[2]);
 
-  // if match_frequency is true, all we need is a direction for k
-  // and a crude guess for its value; we must supply this if k==0.
-  if (match_frequency && k[0] == 0 && k[1] == 0 && k[2] == 0) {
-    k[d-X] = omega_src * sqrt(get_eps(eig_vol.center()));
-    if(!quiet && verbose)
-     master_printf("NEW KPOINT: %g, %g, %g\n", k[0], k[1], k[2]);
-    if (s[d-X] > 0) {
-      k[d-X] *= s[d-X]; // put k in G basis (inverted when we compute kcart)
-      if (fabs(k[d-X]) > 0.4)  // ensure k is well inside the Brillouin zone
-	k[d-X] = k[d-X] > 0 ? 0.4 : -0.4;
-    if(!quiet && verbose)
-      master_printf("NEWER KPOINT: %g, %g, %g\n", k[0], k[1], k[2]);
-    }
-  }
-
   for (int i = 0; i < 3; ++i) {
     n[i] = int(resolution * s[i] + 0.5); if (n[i] == 0) n[i] = 1;
     R[i][i] = s[i] = s[i] == 0 ? 1 : s[i];
     G[i][i] = 1 / R[i][i]; // recip. latt. vectors / 2 pi
   }
 
-  for (int i = 0; i < 3; ++i)
-    for (int j = 0; j < 3; ++j)
-      kcart[i] += G[j][i] * k[j];
-  double klen0 = sqrt(k[0]*k[0]+k[1]*k[1]+k[2]*k[2]);
-  double klen = sqrt(kcart[0]*kcart[0]+kcart[1]*kcart[1]+kcart[2]*kcart[2]);
-  if (klen == 0.0) {
-    if (match_frequency) abort("need nonzero kpoint guess to match frequency");
-    klen = 1;
-  }
-  kdir[0] = kcart[0] / klen;
-  kdir[1] = kcart[1] / klen;
-  kdir[2] = kcart[2] / klen;
-
   maxwell_data *mdata = create_maxwell_data(n[0], n[1], n[2],
 					    &local_N, &N_start, &alloc_N,
 					    band_num, band_num);
   if (local_N != n[0] * n[1] * n[2])
     abort("MPI version of MPB library not supported");
 
+  meep_mpb_eps_data eps_data;
+  eps_data.s = s; eps_data.o = o; eps_data.dim = gv.dim; eps_data.f = this;
+  set_maxwell_dielectric(mdata, mesh_size, R, G, meep_mpb_eps,NULL, &eps_data);
+  if (check_maxwell_dielectric(mdata, 0))
+    abort("invalid dielectric function for MPB");
+
+  double kmatch = G[d-X][d-X] * k[d-X]; // k[d] in cartesian
+  kdir[d-X] = 1; // kdir = unit vector in d direction
+
+  // if match_frequency is true, we need at least a crude guess for kmatch;
+  // which we automatically pick if kmatch == 0.
+  if (match_frequency && kmatch == 0) {
+    vec cen = eig_vol.center();
+    kmatch = omega_src * sqrt(get_eps(cen)*get_mu(cen));
+    k[d-X] = kmatch * R[d-X][d-X]; // convert to reciprocal basis
```

Comment:
This line seems problematic and may be the cause of the incorrect mode reported in #2107.

The fix: `kmatch` should be changed to `(kmatch - kpara_sum)` where `kpara_sum` is the sum of the magnitude of the parallel components of `k` (i.e., parallel to the monitor). Instead, this line assigns a *single* component (in the normal direction `d`) of `k` to be `kmatch` (in Cartesian coordinates). The resulting `k` therefore has the wrong magnitude if any of its other components are non-zero. This may explain why the [mode found by MPB is not what is expected](https://github.com/NanoComp/meep/pull/2107#discussion_r966364983).

Reply:
It's just an initial guess for the Newton iteration; it doesn't have to be that accurate.